### DISCUSSION TITLE
How are option ID's generated?

### DISCUSSION THREAD

**Discussion by JoeMatkin** [swellstores/community#329] [Help] [answered]
We need to know how option ID's are generated as we create an option in our CMS and then send that option to Swell so it can create the variants and then we receive the variants back from Swell and put them into our CMS.

If we send just the option name to Swell we get some issues so we are thinking it's due to us not having an ID and just sending the name. We tried UUID library to generate id's and although it fixed some issue it created others, thanks.

`{
  "options": [
    {
      "id": "What library to generate this?",
      "name": "metals",
      "values": [
        {
          "id": "What library to generate this?",
          "name": "Platinum"
        }
      ],
      "variant": true,
      "input_type": "select"
    },
    {
      "id": "What library to generate this?",
      "name": "stoneShapes",
      "values": [
        {
          "id": "What library to generate this?",
          "name": "Round"
        }
      ],
      "variant": true,
      "input_type": "select"
    }
  ]
}`

**Comment by ericingram** [comment, marked as answer]
@JoeMatkin we use Mongo Object IDs (data type is `objectid` in models)

You can use this npm library to generate them: https://www.npmjs.com/package/bson-objectid. Note creating any record or object will always accept a custom object ID if supplied in the request

**Reply by JoeMatkin** [reply]
Thanks for this, I see.

We have created a custom sync function between sanity and swell, everything is working well except from the operation shown bellow.

https://www.loom.com/share/1214792cc4674de385a53eef52b97cf1
The first examples are it working, the last example are me replacing a single option with another option and this is where we get the problem.

<img width="1221" alt="Screenshot 2023-08-02 at 20 40 33" src="https://github.com/swellstores/community/assets/72705313/5ead7634-0aa9-477e-9d8c-9a438f216885">
After the last example in the video this is what it looks like in swell, this is correct but we think the problem is the variant created webhook hasn't been triggerd therefore changes aren't reflected in Sanity


We send the same request body as Swell sends it through the GUI but for the id's we have tried using UUID to generate the ID which does fix this issue but causes others and we have tried no ID (shown in the above video)

thanks

**Reply by ericingram** [reply]
Hmm is it possible you're just updating an existing variant (i.e. an element in the option array) and so it doesn't trigger `variant.created`, but it would trigger `variant.updated`?

**Reply by JoeMatkin** [reply]
> Hmm is it possible you're just updating an existing variant (i.e. an element in the option array) and so it doesn't trigger `variant.created`, but it would trigger `variant.updated`?

I don't think that's the case, I'll provide an example

Existing variant:
Metal: Rose Gold
Stone Shape: Round

Remove option "Rose Gold"
Add option "Platinum"

Publish in Sanity CMS

Doesn't create webhook but changes in Swell

**Reply by JoeMatkin** [reply]
Ahhhh this is spot on, the options are different so I'm not sure why it's reusing the same variant but it does reuse (using the same variant id)

So as you say it triggers the variant.update webhook

But I still don't understand why the variant ID is being reused, the variant options are different

**Reply by ericingram** [reply]
@JoeMatkin I think it's because variants have a field called `option_value_ids` which tell the variant which options specifically make up the variant. So if the options change (for example it could just be fixing a typo) then it's still linked to the same variant.

You could try to ensure the options are new, not using the same ID, by using $set

**Reply by JoeMatkin** [reply]
Hi @ericingram most of our issue during building the sync functions where caused by us using alternative ID generators instead of BSON object ID as you suggested.

I think it would be helpful to add to the docs that the ID in options should be generated by a BSON object ID otherwise you will get errors such as option.id being null, at least this is what we experienced.

Thanks

**Reply by ericingram** [reply]
Good idea @JoeMatkin - we'll consider adding it
